File: backend/routers/applications.py
from fastapi import APIRouter, HTTPException, Body
from pydantic import BaseModel
from bson import ObjectId
from services.db import get_users_collection, get_resumes_collection, get_jobs_collection, get_matches_collection
from services.gemini import analyze as analyze_with_gemini
from services.scorer import calculate_score
import datetime

router = APIRouter()

class ApplyRequest(BaseModel):
    userId: str
    jobId: str

@router.post("/apply")
async def apply_job(payload: ApplyRequest):
    users = get_users_collection()
    resumes = get_resumes_collection()
    jobs = get_jobs_collection()
    candidates = get_candidates_collection()

    # 1. Fetch Student
    user = await users.find_one({"_id": ObjectId(payload.userId)})
    if not user:
        raise HTTPException(status_code=404, detail="User not found")

    # 2. Fetch Latest Resume
    resume = await resumes.find_one({"userId": ObjectId(payload.userId)})
    if not resume:
        raise HTTPException(status_code=400, detail="No resume found for this student. Please upload one first.")

    # 3. Fetch Job
    job = await jobs.find_one({"_id": ObjectId(payload.jobId)})
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")

    # 4. Run AI Analysis
    resume_text = resume.get("parsedText", "")
    jd_text = job.get("description", "")
    
    try:
        # Use Gemini service for analysis
        analysis = analyze_with_gemini(resume_text, jd_text)
        
        # Calculate scores using the scorer service
        # In a real app, we might want to get semantic similarity from an embedder
        # For now, we'll use a placeholder or check if gemini provided one
        semantic_sim = 0.75 # Placeholder until we integrate embedder if needed
        scores = calculate_score(semantic_sim, analysis)
        analysis.update(scores)
    except Exception as e:
        analysis = {"error": str(e), "ats_score": 0}

    # 5. Save to Matches (Used by Recruiter Dashboard)
    # The Node.js ranking endpoint expects records in the 'matches' collection
    match_record = {
        "resumeId": resume["_id"],
        "jobId": job["_id"],
        "matchScore": int(analysis.get("ats_score", 0)),
        "semanticScore": float(analysis.get("semantic_score", 0.75)),
        "shortlist": False,
        "invited": False,
        "createdAt": datetime.datetime.utcnow()
    }

    matches_col = get_matches_collection()
    res = await matches_col.insert_one(match_record)
    
    return {
        "status": "ok", 
        "match_id": str(res.inserted_id),
        "analysis": analysis
    }

File: backend_node/routes/applications.js
const express = require('express');
const router = express.Router();
const Match = require('../models/Match');
const Resume = require('../models/Resume');

/**
 * Apply for a job
 * Logic: User clicks apply -> find their latest resume -> create Match record
 */
router.post('/apply', async (req, res) => {
  try {
    const { userId, jobId } = req.body;

    if (!userId || !jobId) {
      return res.status(400).json({ error: 'userId and jobId are required' });
    }

    // 1. Find the latest resume for the user
    const latestResume = await Resume.findOne({ userId }).sort({ createdAt: -1 });
    
    if (!latestResume) {
      return res.status(404).json({ error: 'No resume found. Please upload a resume before applying.' });
    }

    // 2. Check if already applied (already exists in Match)
    const existingMatch = await Match.findOne({ jobId, resumeId: latestResume._id });
    if (existingMatch) {
      return res.json({ status: 'ok', message: 'Already applied', match: existingMatch });
    }

    // 3. Create a Match record
    const match = new Match({
      jobId,
      resumeId: latestResume._id,
      matchScore: 0, // Initial score before AI analysis
      semanticScore: 0,
    });

    await match.save();

    res.json({ status: 'ok', message: 'Application submitted successfully', match });
  } catch (err) {
    console.error('Apply Error:', err);
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;
